### PR TITLE
CCDF, CDF, Histogram plugins

### PR DESCRIPTION
**Pull request by Axel-Jacobsen**
Plugins to do some post-processing on recorded data. Depends on [this pull request](https://github.com/pyqtgraph/pyqtgraph/pull/1010) - it looks like pull requests in that repo are being merged relatively slowly, so it may be worth it to implement the changes in jetperch/pyqtgraph in the meantime.

### CODE REVIEW

File: joulescope_ui/range_tool.py [hunk 1]
```diff
@@ -117,6 +117,7 @@ def iterate(self, samples_per_iteration=None):
             progress = self._worker.sigProgress.emit
         else:
             progress = self.progress
+        # Axel Jacobsen, AUG 1 added cast to int for samples_per_iteration to avoid floating point error on range
```
Comment (Axel-Jacobsen) [inline joulescope_ui/range_tool.py]: There must be a better place to solve the issue (i.e. `samples_per_iteration` is a float when it should be an int)

Comment (mliberty1) [reply joulescope_ui/range_tool.py]: Agreed.  The RangeToolIterable.__init__ should enforce the cast, and we should compute it as an int to be clear.  Feel free to fix this, or I can take a look.

**Comment by mliberty1** [comment]
I just tried to run, and I got this:
```
Traceback (most recent call last):
  File "D:\repos\third-party\pyjoulescope_ui_axel\joulescope_ui\range_tool.py", line 40, in run
    msg = self._invoker.range_tool_obj.run(self._invoker)
  File "D:\repos\third-party\pyjoulescope_ui_axel\joulescope_ui\plugins\cdf.py", line 45, in run
    self.hist, bin_edges = histogram_helpers.cdf(data, signal)
  File "D:\repos\third-party\pyjoulescope_ui_axel\joulescope_ui\plugins\histogram_helpers.py", line 42, in cdf
    hist, bin_edges = histogram_helpers.calculate_histogram(data, bins=0, signal=signal)
  File "D:\repos\third-party\pyjoulescope_ui_axel\joulescope_ui\plugins\histogram_helpers.py", line 11, in calculate_histogram
    stats = data.view.statistics_get(t0, t1)['signals'][signal]['statistics']
AttributeError: 'View' object has no attribute 'statistics_get'
ERROR:2019-08-06 20:13:30,413:cdf.py:50:joulescope_ui.plugins.cdf:CDF data is empty
WARNING:2019-08-06 20:13:30,414:main.py:1273:joulescope_ui.main:Error during range tool run()

```
Although we can't rely on view being present, we can add a pre-computed statistics field over the window with mean, variance, min and max fields.

File: joulescope_ui/range_tool.py (hunk 1 above)
Comment (Axel-Jacobsen) [reply joulescope_ui/range_tool.py]: Sounds good! I will fix this and push the update soon.

**Comment by Axel-Jacobsen** [comment]
That would be great! How would one go about adding those pre-computed statistics? And when would `view` be present in `data`?

**Comment by mliberty1** [comment]
Hi @Axel-Jacobsen - I'll take a look this weekend.

**Comment by mliberty1** [comment]
I merged your changes into the "feature/plugins_axel" branch which you can find [here](https://github.com/jetperch/pyjoulescope_ui/tree/feature/plugins_axel).  I also made a number of changes to integrate the plugins cleanly and made some changes to your code, including moving the ".ui" files to the plugins directory.  I also added the project Apache 2.0 license file to the top.  As part of making a contribution, you agree to release these changes under this license.  Take a look and let me know if you have any feedback.

Thanks for your contribution!  I'm really liking the histogram and CDF plots!

**Comment by Axel-Jacobsen** [comment]
Hey good to hear! I like the changes! Is there anything else that you would like me to do for those?

I am also looking at implementing a "max window" functionality which will create new dual markers around the maximum current, voltage, power sum between the original dual markers. I was looking through the code and it seems like to create the new markers, one would need to access an instance of the main windows oscilloscope. Is this correct? What would need to be done to implement this?

Thanks!

**Comment by mliberty1** [comment]
I just merged your changes (along with my modifications) into feature/plugins and then into develop. I also deleted the now unused feature/plugins_axel and feature/plugins branches.  

I also just checked in a change that adds RangeToolInvocation methods marker_single_add and marker_dual_add, which are documented in the [api](https://github.com/jetperch/pyjoulescope_ui/blob/5ffca636f3092ace307bc6b508e0037f2c168bd3/joulescope_ui/plugins/api.py#L67).  See plugins/example_02.py for an example on how to use them.  Let me know if you have any questions!

**Pull request merged**